web/concrete/src/Attribute/Key/CollectionKey.php

```diff
@@ -38,104 +41,122 @@ public static function getAttributes($cID, $cvID, $method = 'getValue') {
-			'cID' => $nvc->getCollectionID(),
-			'cvID' => $nvc->getVersionID(),
-			'akID' => $this->getAttributeKeyID(),
-			'avID' => $av->getAttributeValueID()
-		), array('cID', 'cvID', 'akID'));
-		unset($av);
-	}
-
-	public static function add($at, $args, $pkg = false) {
-
-		// legacy check
-		$fargs = func_get_args();
-		if (count($fargs) >= 5) {
-			$at = $fargs[4];
-			$pkg = false;
-			$args = array('akHandle' => $fargs[0], 'akName' => $fargs[1], 'akIsSearchable' => $fargs[2]);
-		}
-
-		CacheLocal::delete('collection_attribute_key_by_handle', $args['akHandle']);
-
-		$ak = parent::add('collection', $at, $args, $pkg);
-		return $ak;
-	}
-
-	public function delete() {
-		parent::delete();
-		$db = Loader::db();
-		$r = $db->Execute('select avID from CollectionAttributeValues where akID = ?', array($this->getAttributeKeyID()));
-		while ($row = $r->FetchRow()) {
-			$db->Execute('delete from AttributeValues where avID = ?', array($row['avID']));
-		}
-		$db->Execute('delete from CollectionAttributeValues where akID = ?', array($this->getAttributeKeyID()));
-	}
+    public static function getColumnHeaderList()
+    {
+        return parent::getList('collection', array('akIsColumnHeader' => 1));
+    }
+
+    public static function getSearchableIndexedList()
+    {
+        return parent::getList('collection', array('akIsSearchableIndexed' => 1));
+    }
+
+    public static function getSearchableList()
+    {
+        return parent::getList('collection', array('akIsSearchable' => 1));
+    }
+
+    public function getAttributeValue($avID, $method = 'getValue')
+    {
+        $av = CollectionAttributeValue::getByID($avID);
+        if (is_object($av)) {
+            $av->setAttributeKey($this);
+            $value = $av->{$method}();
+            $av->__destruct();
+            unset($av);
+            return $value;
+        }
+    }
+
+    public static function getByID($akID)
+    {
+        $ak = new static();
+        $ak->load($akID);
+        if ($ak->getAttributeKeyID() > 0) {
+            return $ak;
+        }
+    }
+
+    public static function getByHandle($akHandle)
+    {
+        $ak = CacheLocal::getEntry('collection_attribute_key_by_handle', $akHandle);
+        if (is_object($ak)) {
+            return $ak;
+        } else if ($ak == -1) {
+            return false;
+        }
+
+        $ak = new static();
+        $ak->load($akHandle, 'akHandle');
+        if ($ak->getAttributeKeyID() < 1) {
+            $ak = -1;
+        }
+
+        CacheLocal::set('collection_attribute_key_by_handle', $akHandle, $ak);
+
+        if ($ak === -1) {
+            return false;
+        }
+
+        return $ak;
+    }
+
+    /**
+     * Get list of attributes for collection category
+     * @param null $akCategoryHandle Ignored; included for legacy support
+     * @param null $filters Ignored; included for legacy support
+     * @return array
+     */
+    public static function getList($akCategoryHandle = null, $filters = null)
```

Comment:
That's essentially what I was proposing. I had some thoughts on this in https://github.com/concrete5/concrete5/issues/2714 , and essentially the problem with doing a straight check of func_num_args() is that it doesn't nicely handle default argument values. If you have something like `AttributeKey::add()`, that used to take 4 args but the 4th had a default value (and since it was `$pkg`, was probably omitted on most invocations already), anyone calling it would be calling it with 3 arguments.

Since `$args` was already assumed to be an array in add(), I think a simple is_array() check is all that's needed there: https://github.com/jkoudys/concrete5/commit/1a41105b63cbe4241a9cd62fcdaf7b21b9f3fc32#diff-0efc7aa381e4771ae4a2c69528fd7a66R362

I think Korvin's referring to the `getList()` changes. If it's acceptable to just deprecate them, then that sounds ideal to me. I don't really see a beautiful way to do this, but if I made updated Attribute/Key.php like so:

```php
public static function getList()
{
    // Deprecated. getListByCategoryHandle() should be called directly
    list($akCategoryHandle, $filters) = func_get_args();
    return self::getListByCategoryHandle($akCategoryHandle, $filters);
}

public static function getListByCategoryHandle(string $akCategoryHandle, array $filters = array())
{
    // AttributeKey::getList() logic moved to here...
```

On the next version, the getList() behaviour could be replaced with one that just lists all keys (essentially remove the akCategoryHandle from the where clause). I'm not really sure what use getting all keys would be, but it makes sense as the expected result -- if a UserKey::getList() lists all the AttributeKeys in the 'user' category, a Key::getList() should list all AttributeKeys. This keeps things in-line with the LSP. Since we're supporting 5.3+, this could alternatively use late-static bindings: if in AttributeKey object, `static::$akCategoryHandle` is set, use that as the handle in a getList(), otherwise get all keys (or maybe fail?). Obviously none of that will go into this PR, but I'm trying to get a handle on the eventual expected behaviour.